### ISSUE TITLE
broadcom-sta fails to build on 17.09 with latest kernel

### ISSUE THREAD

**Issue by YorikSar** [NixOS/nixpkgs#35323]
## Issue description

When I ran `nixos-rebuild build --upgrade`, it failed to build my configuration because dependencies couldn't be built.

### Steps to reproduce

Relevant sections from `configuration.nix` and `hardware-configuration.nix`:
```nix
{
  boot.kernelPackages = pkgs.linuxPackages_latest;
  boot.extraModulePackages = [ config.boot.kernelPackages.broadcom_sta ];
}
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.15, NixOS, 17.09.2875.c2b668ee726 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-17.09.3031.d09e425aea3"`
 - channels(user): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by teto** [comment]
a full log might be helpful

**Comment by YorikSar** [comment]
@teto Sorry, forgot to attach the log. Here it is:
```
building
build flags: SHELL=/nix/store/jgw8hxx7wzkyhb2dr9hwsd9h2caaasdc-bash-4.4-p12/bin/bash KBASE=/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0
KBUILD_NOPEDANTIC=1 make -C /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/build M=`pwd`
make[1]: Entering directory '/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/build'
building path(s) ‘/nix/store/4pyibnkjymzzrpb3ax36snjcw8qks226-unit-firewall.service’
created 6660 symlinks in user environment
install-info: warning: no info dir entry in `/nix/store/y2r3zy5qw0aqr55dicmpjabmqxm1s5i3-system-path/share/info/time.info'
building path(s) ‘/nix/store/wjijcgrgbbbay2gym402gc34rmrvalqi-dbus-1’
building path(s) ‘/nix/store/bls9m4qymbgvljx43krzd4w66lnfj5qq-unit-polkit.service’
building path(s) ‘/nix/store/fmg7k0j907j8nkz8nnc83kafgn7a4139-unit-systemd-fsck-.service’
building path(s) ‘/nix/store/g5kh87b65lzaa3ykimvii7fnv5gcpy9i-unit-dbus.service’
building path(s) ‘/nix/store/084l87hwimm386aq0rm0s6y931h1rhar-system-units’
building path(s) ‘/nix/store/rnbzl46xsm9n37nc2y250lg449a5pkjc-user-units’
CFG80211 API is prefered for this kernel version
Using CFG80211 API
  CC [M]  /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.o
building path(s) ‘/nix/store/iimapc4anp9b4ar442dnz2wcs8kbny7q-etc’
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c: In function 'osl_os_get_image_block':
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:1083:26: warning: passing argument 2 of 'kernel_read' makes pointer from integer without a cast [-Wint-conversion]
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                          ^~
In file included from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/huge_mm.h:7:0,
                 from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/mm.h:463,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/include/linuxver.h:65,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:25:
/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/fs.h:2842:16: note: expected 'void *' but argument is of type 'loff_t {aka long long int}'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^~~~~~~~~~~
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:1083:37: warning: passing argument 3 of 'kernel_read' makes integer from pointer without a cast [-Wint-conversion]
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                                     ^~~
In file included from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/huge_mm.h:7:0,
                 from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/mm.h:463,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/include/linuxver.h:65,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:25:
/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/fs.h:2842:16: note: expected 'size_t {aka long unsigned int}' but argument is of type 'char *'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^~~~~~~~~~~
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:1083:42: warning: passing argument 4 of 'kernel_read' makes pointer from integer without a cast [-Wint-conversion]
  rdlen = kernel_read(fp, fp->f_pos, buf, len);
                                          ^~~
In file included from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/huge_mm.h:7:0,
                 from /nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/mm.h:463,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/include/linuxver.h:65,
                 from /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/shared/linux_osl.c:25:
/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/include/linux/fs.h:2842:16: note: expected 'loff_t * {aka long long int *}' but argument is of type 'int'
 extern ssize_t kernel_read(struct file *, void *, size_t, loff_t *);
                ^~~~~~~~~~~
  CC [M]  /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.o
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c: In function 'wl_pci_probe':
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c:777:2: warning: this 'if' clause does not guard... [-Wmisleading-indentation]
  if ((val & 0x0000ff00) != 0)
  ^~
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c:779:3: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the 'if'
   bar1_size = pci_resource_len(pdev, 2);
   ^~~~~~~~~
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c: In function 'wl_init_timer':
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c:2360:2: error: implicit declaration of function 'init_timer'; did you mean 'init_timers'? [-Werror=implicit-function-declaration]
  init_timer(&t->timer);
  ^~~~~~~~~~
  init_timers
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c:2361:10: error: 'struct timer_list' has no member named 'data'
  t->timer.data = (ulong) t;
          ^
/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.c:2362:20: error: assignment from incompatible pointer type [-Werror=incompatible-pointer-types]
  t->timer.function = wl_timer;
                    ^
cc1: some warnings being treated as errors
make[4]: *** [/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/scripts/Makefile.build:317: /tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta/src/wl/sys/wl_linux.o] Error
1
make[3]: *** [/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/source/Makefile:1508: _module_/tmp/nix-build-broadcom-sta-6.30.223.271-4.15.drv-0/broadcom-sta] Error 2
make[2]: *** [Makefile:146: sub-make] Error 2
make[1]: *** [Makefile:24: __sub-make] Error 2
make[1]: Leaving directory '/nix/store/p0ci77xjzzhz1j2glzmi8lzn6py21mg1-linux-4.15-dev/lib/modules/4.15.0/build'
make: *** [Makefile:161: all] Error 2
builder for ‘/nix/store/iklkbw44wdpwm3ix0d9y0dk4kc6zp0cm-broadcom-sta-6.30.223.271-4.15.drv’ failed with exit code 2
```

**Comment by YorikSar** [comment]
It seems 65187722e needs to be cherry-picked there...

**Comment by YorikSar** [comment]
I've created a pull request for this: https://github.com/NixOS/nixpkgs/pull/35346

**Comment by fpletz** [comment]
Fixed by #35346
